### PR TITLE
SMT2 uninterpreted functions

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/flattening/boolbv.cpp [hunk 1]
```diff
@@ -565,6 +565,11 @@ literalt boolbvt::convert_rest(const exprt &expr)
     else if(op.type().id() == ID_fixedbv)
       return const_literal(true);
   }
+  else if(expr.id() == ID_function_application)
+  {
+    functions.record(to_function_application_expr(expr));
+    return prop.new_variable();
+  }
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv.cpp:572]: Does this also have implications outside SMT2, i.e., were uninterpreted functions previously broken in some way?

**Review by tautschnig — APPROVED** [review]
Looks ok to me, but I can't claim particular knowledge with the SMT2 back-end or solver.

File: src/solvers/flattening/boolbv.cpp (hunk 1 above)
Comment (martin-cs) [reply src/solvers/flattening/boolbv.cpp:572]: Yeah; I agree with @tautschnig this seems like it should have a visible / testable effect on the usual back-end.

**Review by martin-cs — APPROVED** [review]
Approve modulo @tautschnig 's concern.

File: src/solvers/flattening/boolbv.cpp (hunk 1 above)
Comment (kroening) [reply src/solvers/flattening/boolbv.cpp:572]: > Does this also have implications outside SMT2, i.e., were uninterpreted functions previously broken in some way?

I've added a further test that uses the C front-end.

**Pull request merged**